help/logging-out: Use tabbed navigation for instructions.

eeshangarg opened the pull request:
@rishig: I think we might have to modify the CSS of the content that ends up inside the tabbed panes. Let me know what you think! Thanks! :)

eeshangarg commented:
@rishig: The build failing is due to the casper flake! :)

rishig commented:
cool, thanks! I think it looks pretty good already. I think we can just copy the zendesk styling, e.g. https://get.slack.help/hc/en-us/articles/214613347-Sign-out-of-Slack#desktop-web-1. 

Also, how hard would it be to 
* add navigation, e.g. the `#desktop-web` at the end? For linking from the mobile app, would be nice to be able to just link to the mobile instructions. 
* autodetect your user agent and choose a tab appropriately?
* reduce the amount of text needed in the help file? E.g. via a macro or something in the bugdown processor.

eeshangarg commented:
@rishig: I spent some time reading the code that would allow me to make the changes you requested. 
 Okay, so here are my thoughts on the questions you asked:

* I'm not sure I understand. What do you mean by "add navigation ... at the end?"
* That should not be too hard. I think we already have the requisite code in `static/js/portico/landing-page.js`. It would simply be a matter of activating the right tab when rendering the `landing-page` JS bundle.
* Having written a few Markdown extensions for Zulip, I think this is semi-hard, but definitely doable. The difficult part would be to come up with syntax that doesn't clash with existing extensions and code block syntax and what not! So, something like:
```
~~~ tabbed("installing-on-macos", group=1)
### Instructions go here
~~~
```
where, `group=1` indicates different sections of tabbed instructions, in the case that we end up with a page that has two sets of tabbed instructions or something.

What do you think? Also, which one would you like to see get done first?

Thanks!

rishig commented:
Thanks for looking into it!
1. Sorry, at the end of the URL. So `https://zulipchat.com/help/logging-out#android` would go to the Android tab.
2. Cool! Let's do it, then; will make it easier to add links to the help pages from the mobile app.
3. ok, let's brainstorm on czo. 

> which one would you like to see get done first?

Let's just do them in order of easiness!

rishig commented:
Nice!
* Looking at the code it seems like we can actually do better, with syntax like 
```
{tabbed_section_start}
{tab desktop|Desktop/browser}
desktop content
{tab ios|iOS}
ios content
{tabbed_section_end}
```
Technically we can even do it without the initial `{tabbed_section_start}`, but I think it'll be good to have it for readability :).
This will also allow us to have multiple tabbed sections in an article.
* There are only a small number of possibilities for the tab names. So we can do something like we do in the top of zerver/lib/bugdown/help_settings_links.py. The syntax will then be something like  
```
{tabbed_section_start}
{tab|desktop-web}
desktop content
{tab|ios}
ios content
{tabbed_section_end}
```
* Let's rename `{tabbed_section_start}` to `{start_tabs}` and `{end_tabs}`.

Possibly we'll want to make it more serious looking, like `{!start_tabs!}`, but that's an easy change for later.

As follow-up, we should rename `{settings_tab|foo}` to `{go-to-settings|foo}` or similar so that we don't overload `tab`.

Thanks @eeshangarg!

eeshangarg commented:
@rishig: Thanks for the review! I just updated this:

* The code looks much cleaner now in general. I rewrote most of it. The initial commit was just a WIP to get feedback, the latest changes should be more readable now.
* Also, we can't use `{!start_tabs!}` since the exclamation mark syntax clashes with our `include` Markdown extension.
* This extension should be able to handle multiple tabbed sections in one Markdown file.

Looking forward to your feedback! Thanks!

eeshangarg commented:
@showell: Updated! :)

eeshangarg commented:
@rishig, @showell: I just pushed a couple of more commits to this:

* Added support for detecting the user's OS and activating the right tab automatically.
* Also migrated `help/desktop-app-install-guide.md` to use tabbed instructions.

For the former, please keep in mind that I am not a JavaScript programmer by any means, so if anything I wrote is terrible, please let me know, and I'll fix it right away! Thanks! :)

rishig commented:
Merged, thanks @eeshangarg! This is a huge step forward for the help docs. 

As followup I'll see how we can style these appropriately.
